docs/architecture/adr-047-handling-evidence-from-light-client.md

```diff
@@ -19,161 +20,200 @@ the light client detector receives two conflicting headers
+client side in the event that an attack happens. Although, this is not a significant 
+difference as the light client would in any case have to validate all the headers 
+from both witness and primary. Using traces would consume a large amount of bandwidth
+and adds a ddos vector to the full node.
 
-It then collects all the headers that were fetched as part of bisection into two traces. 
-One containing the primary's headers and the other containing the witness's headers. 
-The light client sends the trace to the opposite provider (i.e. the primary's trace
-to the witness and the witness's trace to the primary) as the light client is 
-incapable of deciding which one is malicious. 
 
-Assuming one of the two providers is honest, that full node will then take the trace
-and extract out the relevant evidence and propogate it until it ends up on chain.
+## Decision
+
+The light client will be divided into two components: a `Verifier` (either sequential or 
+bisection) and a `Detector` (see [Informal's Detector](https://github.com/informalsystems/tendermint-rs/blob/master/docs/spec/lightclient/detection/detection.md))
+. The detector will take the trace of headers from the primary and check it against all 
+witnesses. For a witness with a diverging header, the detector will first verify the header
+by bisecting through all the heights defined by the trace that the primary provided. If valid,
+the light client will trawl through both traces and find the point of bifurcation where it
+can proceed to extract any evidence (as is discussed in detail later). 
 
-*NOTE: We do not send evidence to other connected peers*
+Upon successfully detecting the evidence, the light client will send it to both primary and
+witness before halting. It will not send evidence to other peers nor continue to verify the
+primary's header against any other header.
 
-*NOTE: The light client halts then and does not verify with any other witness*
 
 ## Detailed Design
 
-The traces will have the following data structure: 
+The verification process of the light client will start from a trusted header and use a bisectional
+algorithm to verify up to a header at a given height. This becomes the verified header (does not 
+mean that it is trusted yet). All headers in between are cached and known as intermediary headers.
 
-```go
-type ConflictingHeadersTrace struct {
-  Headers []*types.SignedHeader
-}
+The light client's detector then takes the headers and runs the detect function.
+
+```golang
+Detect(trace []*SignedHeader, witnesses []Provider) ([]Evidence, error)
 ```
 
-When a full node receives a `ConflictingHeadersTrace`, it should
-a) validate it b) figure out if malicious behaviour is obvious (immediately
-slashable) or run the amnesia protocol.
+This queries the header of the last height for each of the witnesses in order to compare hashes. In the event that these two hashes are 
+different we have verified conflicting headers and therefore the detector proceeds to run 
```

Comment:
At this point in time (after hash comparisons) the conflicting header is not yet verified, i.e., it could be just bogus header. Only after we run bisection from a trusted header with that witness, the conflicting header is verified.